### ISSUE TITLE
[Feature Request] Postponed evaluation of self-referential type hints

### ISSUE THREAD

**Issue by patrick-kidger** [beartype/beartype#217]
```python
from beartype import beartype
from typing import Generic, TypeVar

T = TypeVar("T")

class Foo(Generic[T]):
    @beartype
    def bar(self) -> "Foo[T]":
        return self

Foo().bar()
# beartype.roar.BeartypeDecorHintForwardRefException: Forward reference "__main__.Foo[T]" syntactically invalid as module attribute name.
```
:(

**Comment by leycec** [comment]
*...gah.* You can't put anything over on @patrick-kidger. @beartype currently only supports the stricter (and ostensibly lamer) variety of forward references standardized by PEP 484: that is, forward references to a simple type that has yet to be declared.

Even though I'm *pretty* sure that full-blown postponed evaluation of arbitrarily complex stringified type hints was never standardized, the truth is irrelevant! @beartype should do this regardless. If we don't do this, then everyone will just use [PEP 563 (i.e., `from __future__ import annotations`)](https://peps.python.org/pep-0563/) instead. But *nobody* should use PEP 563, because PEP 563 fundamentally breaks Python in dark ways that are profoundly difficult (and costly) to repair at runtime.

That said... this is non-trivial. This isn't infeasible, because @beartype already does something like this to resolve arbitrarily complex type hints implicitly stringified by PEP 563. Still, we are surely enabling *Hard Mode* here. Let's ponder this as I stroke my beard stubble thoughtfully.

**Comment by patrick-kidger** [comment]
What makes this different to the existing PEP563 support?
I opened two separate issues for completeness but I was guessing this was just a different version of #218. (With stringification of only some annotations, rather than stringification of all of them.)

**Comment by leycec** [comment]
:sweat_smile: 

Much apologies for my belated reply. It's been one of *those* weeks. Surprisingly, this feature request, #219, and #222 are all distinct issues. Submitting one issue for each was perfect on your part. We'll iteratively resolve each one until we exhaust ourselves and quietly give up.

Hard things are the things worth doing. Thankfully, this is probably the most important *and* easiest to fix of these issues. If we lightly rewrite your lead example as...

```python
from beartype import beartype
from typing import Generic, TypeVar

T = TypeVar("T")

@beartype
class Foo(Generic[T]):
    def bar(self) -> "Foo[T]":
        return self

Foo().bar()
```

...then `@beartype` already has *all* of the local and global attributes it needs to fully resolve the stringified type hint `"Foo[T]"`. Since `Foo` has already been declared by the time the `@beartype` decorator is invoked, there's no need for `@beartype` to defer the resolution of `"Foo[T]"`. Instead, `@beartype` can just trivially resolve that hint immediately with an `eval()` call and generate the requisite type-checking code.

Let's tentatively slate this for `beartype 0.14.0` – to be released sometime this Summer, surely! Who could doubt my words? We'll obsessively knuckle down on forward references and stringified type hints for that dev cycle.

Thanks yet again for all these stupendous issues, @patrick-kidger. Thumbs are way up. :+1:

**Comment by leycec** [comment]
**Ah-ha!** The recently released @beartype 0.14.0 *effectively* resolves this without *actually* resolving this. This isn't what anyone was expecting or actually wanted. @beartype surprises us all, because...

@beartype now [fully supports PEP 673 (i.e., `typing.Self`)](https://beartype.readthedocs.io/en/latest/faq/#the-current-class) – the new gold standard for doing exactly what you want to do here. `typing.Self` is dramatically preferable to forward references and PEP 563 (i.e., `from __future__ import annotations`). `typing.Self` is a *far* faster, stabler, and more convenient API for type-checking classes self-referentially. We shill for `typing.Self`.

Take it away, minimal reproducible example (MRE)!

```python
from beartype import beartype
from typing import Generic, TypeVar
from typing import Self  # <---------------- if Python ≥ 3.11.0
# from typing_extensions import Self   # <-- if Python < 3.11.0

T = TypeVar("T")

@beartype  # <-- "Self" requires @beartype on classes rather than methods.
class Foo(Generic[T]):
    def bar(self) -> Self:  # <-- That's what I'm talkin' 'bout, yo. Blam!
        return self

Foo().bar()
```

At runtime, @beartype now type-checks `Foo.bar()` as returning an instance of `Foo`. At static analysis time, mypy and friends type-check `Foo.bar()` as returning an object satisfying `Foo[T]`. Everyone is happy. No one is sad. This is surely the road to world peace in our time.

Thanks so much for your venerable patience, @patrick-kidger. @beartype will get you and the JAX community sorted "pronto." <sup>1</sup> :sweat_smile: 

<sup>1. Definitions of "pronto" may vary by open-source supplier.</sup>

**Comment by leycec** [comment]
@beartype summons the ultimate destructive force to deal with this long-standing issue: apocalyptic anime exploding girl, go!

![arise ugh](https://media.tenor.com/V6k3KbrycpsAAAAC/anime-explosion.gif)

**Comment by leycec** [comment]
**Resolved by 4f7115919c...** but actually by a18d9a5a5da, which this issue turned out to be a duplicate of. I shall now summon [my younger self circa last Friday night](https://github.com/beartype/beartype/issues/226#issuecomment-1694160566), who personally witnessed [the tragic chain of events that ultimately lead to @beartype's redemption and this final resolution](https://github.com/beartype/beartype/issues/226#issuecomment-1694160566):

## Younger Self Witnesses All, Now Tells Sad Story (Only on GitHub)

**omg! OMG! OMGGGGG!!!** @TeamSpen210: [Your brilliant five-sentence abstract on the nature of Python vis-a-vis forward references](https://github.com/beartype/beartype/issues/226#issuecomment-1484045900) *actually* worked! It *actually* worked, bro! Punch the air like you just don't care, 'cause that's what we're doin'.

![do it like this](https://media.tenor.com/hV03QzfXrtEAAAAC/hatsune-priconne.gif)

Naturally, your "simpler approach" was horrifying. My once-beautific bald face is crinkled with wrinkles, my twitchy eye is even more twitchy, and our cat won't stop pacing the room. I spent the entirety of my vacation doing this – and I don't even *need* this in production code. It just became a titanic struggle of @leycec versus @beartype versus Python: a three-way untelevised cage match that could only end with one man hurtling fifty feet onto another man lying prostrate on a table perched on top of another man curled in the fetal position while bleeding out live on GitHub.

![mankind is sad](https://media.tenor.com/t-Tw6ma_jxIAAAAC/undertaker-mankind.gif)
<sup>...dat the best u can do, python?</sup>

![wat](https://media.tenor.com/T1E389Up3RgAAAAC/jeff-hardy-whisper-in-the-wind.gif)
<sup>*wat is even happening here*</sup>

![uh gods](https://media.tenor.com/3U1flZRg0f8AAAAC/adam-cole-johnny-gargano.gif)
<sup>When your masterplan only ends in tragedy for everybody.</sup>

Seriously. I've got five hundred-line comments devsplaining single one-liners. I've got 100KB of incomprehensible pure-Python isolated to four new submodules sprawled across a new subpackage devoted exclusively to inchoate madness. I've got abstract base classes, abstract metaclasses, dynamic type factories, dunder methods on sunder methods on abstract cached properties. I've got friggin' metaclass `__getattr__()` methods performing deferred module attribute importation via dynamic class attribute lookups of attributes that don't actually physically exist anywhere. I've got two whole `dict.__missing__()` implementations that don't even really have anything to do with one another. *This is inchoate madness.*

I don't even know how any of this works anymore. I've already jettisoned all knowledge pertaining to this topic so I can just sleep tonight without fitfully crying out in the night for my childhood stuffed octopus. I wish to awaken from the real-world nightmare I have unwittingly exposed myself to. It turns out *Bloodborne* is more a state of mind rather than a mere video game.

![who knew](https://media.tenor.com/-AMHQLsgQXcAAAAd/bloodborne-werewolf.gif)
<sup>Python. *It do be like that.*</sup>

Forward references: do not go there, lest this animated reenactment of @leycec's life corrupt your once-pure essence.

I'm pretty sure (but not certain) that I may have inadvertently summoned the Eldritch She-God Shub-Niggurath, The Black Goat of the Woods with a Thousand Young, during the resolution of this feature request. I apologize, humanity. I'm so sorry! I didn't *mean* to open that portal. It just opened of its own accord while the guard cat – a fat Maine Coon named Nekali, of course – was indolently passed out on the shag carpeting.
